stdlib/public/core/DictionaryBuilder.swift

```diff
@@ -69,58 +69,102 @@ extension Dictionary {
+        allowingDuplicates: allowingDuplicates,
+        initializingWith: initializer))
+  }
+}
+
+extension _NativeDictionary {
+  @inlinable
+  internal init(
+    _unsafeUninitializedCapacity capacity: Int,
+    allowingDuplicates: Bool,
+    initializingWith initializer: (
+      _ keys: UnsafeMutableBufferPointer<Key>,
+      _ values: UnsafeMutableBufferPointer<Value>,
+      _ initializedCount: inout Int
+    ) -> Void
+  ) {
+    self.init(capacity: capacity)
+    var initializedCount = 0
+    initializer(
+      UnsafeMutableBufferPointer(start: _keys, count: capacity),
+      UnsafeMutableBufferPointer(start: _values, count: capacity),
+      &initializedCount)
     _precondition(count >= 0 && count <= capacity)
+    _storage._count = initializedCount
 
-    // Hash elements into the correct buckets.
-    var bucket = _HashTable.Bucket(offset: 0)
-    while bucket.offset < count {
-      if native.hashTable._isOccupied(bucket) {
+    // Hash initialized elements and move each of them into their correct
+    // buckets.
+    //
+    // - We have some number of unprocessed elements at the start of the
+    //   key/value buffers -- buckets up to and including `bucket`. Everything
+    //   this region is either unprocessed or in use. There are no uninitialized
+    //   entries in it.
+    //
+    // - Everything after `bucket` is either uninitialized or in use. This
+    //   region works exactly like regular dictionary storage.
+    //
+    // - "in use" is tracked by the bitmap in `hashTable`, the same way it would
+    //   be for a working Dictionary.
+    //
+    // Each iteration of the loop below processes an unprocessed element, and/or
+    // reduces the size of the unprocessed region, while ensuring the above
+    // invariants.
+    var bucket = _HashTable.Bucket(offset: initializedCount - 1)
+    while bucket.offset >= 0 {
+      if hashTable._isOccupied(bucket) {
         // We've moved an element here in a previous iteration.
-        bucket.offset += 1
+        bucket.offset -= 1
         continue
       }
-      let hashValue = native.hashValue(for: native._keys[bucket.offset])
-      let target: _HashTable.Bucket
-      if _isDebugAssertConfiguration() {
-        let (b, found) = native.find(
-          native._keys[bucket.offset],
-          hashValue: hashValue)
-        guard !found else {
-          _preconditionFailure("Duplicate keys found")
+      // Find the target bucket for this entry and mark it as in use.
+      let target: Bucket
+      if _isDebugAssertConfiguration() || allowingDuplicates {
+        let (b, found) = find(_keys[bucket.offset])
+        if found {
+          _internalInvariant(b != bucket)
+          _precondition(allowingDuplicates, "Duplicate keys found")
+          // Discard existing entry, then move the current entry in place of it.
+          uncheckedDestroy(at: b)
+          _storage._count -= 1
+          moveEntry(from: bucket, to: b)
```

Comment:
Wouldn't it be cheaper to leave the existing entry where it is? Why does the first one win?

Reply:
This is how the existing bridging code works for String keys. Arguably the behavior is observable -- although you have to *really* work for it, so it probably isn't worth worrying about it.

Reply:
Ah, this is already breaking first/last ordering anyway because of all the swapping it does. I'll change it to discard the current entry.